community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArraySet.java

```diff
@@ -0,0 +1,334 @@
+ */
+package org.neo4j.cypher.internal.runtime;
+
+/**
+ * When you need to have a set of arrays of longs representing entities - look no further
+ * <p>
+ * This set will keep all it's state in a single long[] array, marking unused slots
+ * using 0xF000000000000000L, a value that should never be used for node or relationship id's.
+ * <p>
+ * The set will be resized when either probing has to go on for too long when doing inserts,
+ * or the load factor reaches 75%.
+ *
+ * The word "offset" here means the index into an array,
+ * and slot is a number that multiplied by the width of the values will return the offset.
+ */
+public class LongArraySet
+{
+    private static final long NOT_IN_USE = -2;
+
+    private static final int SLOT_EMPTY = 0;
+    private static final int VALUE_FOUND = 1;
+    private static final int CONTINUE_PROBING = -1;
+
+    private Table table;
+    private final int width;
+
+    public LongArraySet( int initialCapacity, int width )
+    {
+        assert (initialCapacity & (initialCapacity - 1)) == 0 : "Size must be a power of 2";
+        assert width > 0 : "Number of elements must be larger than 0";
+
+        this.width = width;
+        table = new Table( initialCapacity );
+    }
+
+    /**
+     * Adds a value to the set.
+     *
+     * @param value The new value to be added to the set
+     * @return The method returns true if the value was added and false if it already existed in the set.
+     */
+    public boolean add( long[] value )
+    {
+        assert value.length == width : "all elements must have the same size";
+        int slotNr = slotFor( value );
+        while ( true )
+        {
+            int offset = slotNr * width;
+            if ( table.inner[offset] == NOT_IN_USE )
+            {
+                if ( table.timeToResize() )
+                {
+                    // We know we need to add the value to the set, but there is no space left
+                    resize();
+                    // Need to restart linear probe after resizing
+                    slotNr = slotFor( value );
+                }
+                else
+                {
+                    table.setValue( slotNr, value );
+                    return true;
+                }
+            }
+            else
+            {
+                for ( int i = 0; i < width; i++ )
+                {
+                    if ( table.inner[offset + i] != value[i] )
+                    {
+                        // Found a different value in this slot
+                        slotNr = (slotNr + 1) & table.tableMask;
+                        break;
+                    }
+                    else if ( i == width - 1 )
+                    {
+                        return false;
+                    }
+                }
+            }
+        }
+    }
+
+    /***
+     * Returns true if the value is in the set.
+     * @param value The value to check for
+    1     * @return whether the value is in the set or not.
```

Comment:
`1`?

Reply:
Remind me not to program with buttered fingers again, please.